bad-bits/lib/store.js

```diff
@@ -1,58 +1,130 @@
+const BAD_BITS_CID_LENGTH = 64
+const KV_SEGMENT_MAX_HASH_COUNT = Math.floor(
+  KV_VALUE_MAX_SIZE / (BAD_BITS_CID_LENGTH + 1) /* , */,
+)
+const MAX_TOTAL_CHANGES = 10_000
+const MAX_KV_BATCH_SIZE = 1_000
+
 /**
  * Updates the bad bits database with new hashes
  *
- * @param {Env} env - Environment containing database connection
+ * @param {{ BAD_BITS_KV: KVNamespace }} env - Environment containing database
+ *   connection
  * @param {Set<string>} currentHashes - Set of current valid hashes from
  *   denylist
  * @param {string | null} etag - ETag for the current denylist
  */
 export async function updateBadBitsDatabase(env, currentHashes, etag) {
   const startedAt = Date.now()
-  const timestamp = new Date().toISOString()
-  const insertBadBitStmt = env.DB.prepare(
-    `
-    INSERT INTO bad_bits (hash, last_modified_at) VALUES (?, ?)
-    ON CONFLICT(hash) DO UPDATE SET last_modified_at = excluded.last_modified_at
-    `,
-  )
 
-  let updated = 0
-  const remainingHashes = Array.from(currentHashes)
-  while (remainingHashes.length > 0) {
-    // pop first N hashes from remainingHashes
-    const batchHashes = remainingHashes.splice(0, 10_000)
+  console.log('getting latest hashes')
+  const oldHashes = new Set(await getAllBadBitHashes(env))
 
-    await env.DB.batch(
-      batchHashes.map((hash) => insertBadBitStmt.bind(hash, timestamp)),
-    )
+  console.log('comparing current hashes')
+  const addedHashes = [...currentHashes.difference(oldHashes)]
+  const removedHashes = [...oldHashes.difference(currentHashes)]
+  addedHashes.splice(MAX_TOTAL_CHANGES)
+  removedHashes.splice(MAX_TOTAL_CHANGES - addedHashes.length)
 
-    updated += batchHashes.length
-    console.log(
-      `Inserted/updated ${updated} bad bits in ${Date.now() - startedAt}ms`,
-    )
+  await persistUpdates(env, oldHashes, addedHashes, removedHashes)
+
+  const wasCapped =
+    addedHashes.length + removedHashes.length >= MAX_TOTAL_CHANGES
+  if (etag && !wasCapped) {
+    await env.BAD_BITS_KV.put('latest-etag', etag)
   }
 
-  const statements = [
-    env.DB.prepare('DELETE FROM bad_bits WHERE last_modified_at < ?').bind(
-      timestamp,
-    ),
-  ]
+  console.log(
+    `+${addedHashes.length} -${removedHashes.length} hashes in ${Date.now() - startedAt}ms`,
+  )
+}
 
-  if (etag) {
-    statements.push(
-      env.DB.prepare(
-        'INSERT INTO bad_bits_history (timestamp, etag) VALUES (?, ?)',
-      ).bind(timestamp, etag),
+/** @param {{ BAD_BITS_KV: KVNamespace }} env */
+export async function getLastEtag(env) {
+  return await env.BAD_BITS_KV.get('latest-etag')
+}
+
+/**
+ * Gets all bad bit hashes from the database
+ *
+ * @param {{ BAD_BITS_KV: KVNamespace }} env - Environment containing database
+ *   connection
+ * @returns {Promise<string[]>} - Array of hash strings
+ */
+export async function getAllBadBitHashes(env) {
+  const hashes = []
+  const { keys } = await env.BAD_BITS_KV.list({ prefix: 'latest-hashes' })
```

Comment:
```suggestion
  const { keys } = await env.BAD_BITS_KV.list({ prefix: 'latest-hashes:' })
```

For consistency with line 123 below.

(I personally prefer to include the delimiter `:` in the filter, to avoid subtle error if we introduce another prefix sharing the initial string, e.g. `latest-hashes-count`.)